tap_klaviyo/utils.py

```diff
@@ -179,68 +179,113 @@ def authed_get(source, url, params):
+                {
+                    "filter": "equals(subscriptions.email.marketing.suppression.reason,'USER_SUPPRESSED')",
+                    "additional-fields[profile]": "subscriptions,predictive_analytics"
+                },
+                {
+                    "filter": "equals(subscriptions.email.marketing.suppression.reason,'UNSUBSCRIBE')",
+                    "additional-fields[profile]": "subscriptions,predictive_analytics"
+                },
+                {
+                    "filter": "equals(subscriptions.email.marketing.suppression.reason,'INVALID_EMAIL')",
+                    "additional-fields[profile]": "subscriptions,predictive_analytics"
+                }
+            ]
+        else:
+            # params list for lists stream 
+            params_list = [
+                {
+                    "include": "tags"
+                }
+            ]
+
+        for params in params_list:
+            for response in get_all_using_next(resource['stream'], endpoint, headers, params):
+                records = response.json().get('data')
+                included = response.json().get('included', [])
+                counter.increment(len(records))
+                transfrom_and_write_records(records, resource, included, params.get("include","").split(","))
+
+
+def transfrom_and_write_records(events, stream, included, valid_relationships):
     event_stream = stream['stream']
     event_schema = stream['schema']
     event_mdata = metadata.to_map(stream['metadata'])
 
     with Transformer() as transformer:
         for event in events:
+            # Flatten the event dict with attributes
+            event.update(event['attributes'])
+            for relationship_key, relationship_value in event.get('relationships',{}).items():
+                if not relationship_key in valid_relationships:
+                    continue
+                relationship_data = relationship_value['data']
+                # Generalizing relationship data to list of dicts for all streams
+                # This is due to the fact that, for Full table streams, data is returned as a list
+                # And, for incremental streams, data is return as a dict in API response
+                if isinstance(relationship_data, dict):
+                    relationship_data = [relationship_data]
+                for relationship in relationship_data:
+                    # Flatten the relationship dict with data
+                    relationship_id = relationship['id']
+                    for included_relationship in included:
+                        # Break out of the loop if included relationship is found
+                        if relationship_id == included_relationship['id']:
+                            # Flatten the included_relationship dict with attributes
+                            included_relationship.update(included_relationship['attributes'])
+                            relationship.update(included_relationship)
```

Comment:
use a mapping / dict to fetch the related relationship in the included list, as going through the whole list for each event will be time consuming.

Reply:
Done